AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1446,16 +1432,20 @@ - (void)__layout
 {
   ASDisplayNodeAssertMainThread();
   ASDN::MutexLocker l(__instanceLock__);
-  CGRect bounds = self.bounds;
-
-  [self measureNodeWithBoundsIfNecessary:bounds];
-
+  CGRect bounds = _threadSafeBounds;
+  
   if (CGRectEqualToRect(bounds, CGRectZero)) {
     // Performing layout on a zero-bounds view often results in frame calculations
     // with negative sizes after applying margins, which will cause
     // measureWithSizeRange: on subnodes to assert.
     return;
   }
+
+  // This method will confirm that the layout is up to date (and update if needed).
+  // Importantly, it will also APPLY the layout to all of our subnodes if there is not already another layout
+  // transition in progress.
+  [self _locked_measureNodeWithBoundsIfNecessary:bounds];
```

Comment:
Honestly I think we should just put the whole implementation of this method here in __layout.  Thoughts:

- It looks like the zero-bounds check is duplicated in the measure method anyway.
- It's really hard to describe what this measure method is doing -- it really does do "check if layout is valid, update it if not, and apply it to all subnodes BUT ONLY IF not transitioning!"  This is a lot for one method to do, but makes more sense as the overall process of the __layout method itself rather than a sub-method.

We could factor out the few lines dedicated to placeholder management if desired (updatePlaceholderLayout or something), but not essential.  So the method wouldn't be much longer.

I think longer term when we come back to this code, we can then break up the implementation of __layout into a couple different stages -- like validate layout, re-measure if necessary, layout application, etc.